pmda/density.py

```diff
@@ -0,0 +1,222 @@
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(gridcenter, xdim, ydim, zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicate when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - padding
+            smax = np.max(coord, axis=0) + padding
+            BINS = fixedwidth_bins(delta, smin, smax)
+            arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+            bins = BINS['Nbins']
+            # create empty grid with the right dimensions (and get the edges)
+            grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins, range=arange, normed=False)
+            grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        h = grid.copy()
+        self._arange = arange
+        self._bins = bins
+
+    def _prepare(self):
+        """
+        Additional preparation to run
+        """
+
+        # pm = ProgressMeter(self._n_frames, interval=interval,
+        #                    verbose=verbose,
+        #                    format="Histogramming %(n_atoms)6d atoms in frame "
+        #                    "%(step)5d/%(numsteps)d  [%(percentage)5.1f%%]\r")
+        # start, stop, step = u.trajectory.check_slice_indices(start, stop, step)
+
+    def _single_frame(self, ts, atomgroups):
+        """
+        Performs computation on single trajectory frame.
+
+        Creates a histogram of positions of atoms in an atom group for a single frame.
+
+        Parameters
+        ----------
+        ts : int
+            current time step
+        atomgroups : AtomGroup
+            atom group for current block
+
+        Returns
+        -------
+
+        """
+        coord = self.current_coordinates(atomgroups[0], self._cutoff, self._atomselection, self._soluteselection, self._update_selection)
+        h, edges = np.histogramdd(coord, bins=self._bins, range=self._arange, normed=False)
+        h = np.array(h)
+        return [h, edges]
+
+    def _conclude(self):
+        """
+        """
+        self._edges = self._results[0, 1]
+        self._grid = self._results[:, 0].sum(axis=0)
+        self._grid /= float(self._n_frames)
+        metadata = self._metadata if self._metadata is not None else {}
+        # metadata['psf'] = u.filename
```

Comment:
You can use `self._atomgroup.universe.filename` (or `self._atomgroup.universe.topology.filename` ?)